sys/fido2/ctap/ctap.c

```diff
@@ -0,0 +1,1508 @@
+                                 bool up, uint32_t sign_count);
+/**
+ * @brief Initialize authData of assertion object
+ */
+static int make_auth_data_next_assert(uint8_t *rp_id_hash,
+                                      ctap_auth_data_header_t *auth_data,
+                                      bool uv, bool up, uint32_t sign_count);
+/**
+ * @brief Find most recent rk matching rp_id_hash and present in allow_list
+ */
+static uint8_t find_matching_rks(ctap_resident_key_t *rks, size_t rks_len,
+                                 ctap_cred_desc_alt_t *allow_list,
+                                 size_t allow_list_len,
+                                 uint8_t *rp_id, size_t rp_id_len);
+/**
+ * @brief Check if any of the credentials in li belong to this authenticator
+ */
+static bool rks_exist(ctap_cred_desc_alt_t *li, size_t len, uint8_t *rp_id,
+                      size_t rp_id_len);
+
+/**
+ * @brief Decrypt credential that is stored by relying party
+ *
+ * Webauthn specification (version 20190304) section 4 (Credential ID)
+ */
+static int ctap_decrypt_rk(ctap_resident_key_t *rk, ctap_cred_id_t *id);
+
+static int save_rk(ctap_resident_key_t *rk);
+static int save_pin(uint8_t *pin, size_t len);
+static int save_state(const ctap_state_t *state);
+static void load_state(ctap_state_t *state);
+static inline bool pin_protocol_supported(uint8_t version);
+static inline uint8_t get_remaining_pin_attempts(void);
+static int decrement_pin_attempts(void);
+static void reset_pin_attempts(void);
+static void reset(void);
+static int verify_pin_auth(uint8_t *auth, uint8_t *hash, size_t len);
+static inline bool locked(void);
+static inline bool boot_locked(void);
+static inline uint16_t get_page_offset(uint16_t i);
+static inline uint16_t get_offset_into_page(uint16_t t);
+
+static ctap_state_t g_state;
+static ctap_get_assertion_state_t g_assert_state;
+static uint8_t g_pin_token[CTAP_PIN_TOKEN_SIZE];
+/* remaining PIN attempts until authenticator is boot locked */
+static int g_rem_pin_att_boot = CTAP_PIN_MAX_ATTS_BOOT;
+
+int fido2_ctap_init(void)
+{
+    int ret;
+
+    load_state(&g_state);
+
+    ret = fido2_ctap_crypto_init();
+
+    if (ret != CTAP2_OK) {
+        DEBUG("fido2_ctap: initialization failed \n");
+        return -1;
+    }
+
+    /* first startup up of the device */
+    if (g_state.initialized != CTAP_INITIALIZED_MARKER) {
+        reset();
+    }
+
+    /* initialize pin_token */
+    fido2_ctap_crypto_prng(g_pin_token, sizeof(g_pin_token));
+
+    DEBUG("fido2_ctap: initialization successful \n");
+
+    return 0;
+}
+
+static void reset(void)
+{
+    g_state.initialized = CTAP_INITIALIZED_MARKER;
+    g_state.rem_pin_att = CTAP_PIN_MAX_ATTS;
+    g_state.pin_is_set = false;
+    g_state.rk_amount_stored = 0;
+    g_state.sign_count = 0;
+
+    g_rem_pin_att_boot = CTAP_PIN_MAX_ATTS_BOOT;
+
+    fido2_ctap_crypto_prng(g_state.cred_key, sizeof(g_state.cred_key));
```

Comment:
This creates a new AES_CCM key to encrypt credentials when they are not stored on the authenticator. I could move this line into `fido2_ctap_encrypt_rk` but this would require additional logic to check if `g_state.cred_key` is initialized or not.

Reply:
> Short notice: I've tested again with the nrf52840dk.
> 
>     1. webauthn.io still works like charm. I was wondering, what is the maximum number of accounts per authenticator with your implementation? So far, i've only tested with two accounts.
> 
>     2. The unit test seems to work in general. Happy to see that you were able to move to the original repo. by now, they almost doubled the number of tests +1
>        There are, however, still some errors. Would you mind to take a look at
> 
> 
> this output
> 
> ```
> ========================================================================= FAILURES ==========================================================================
> _______________________________________________________________________ test_lockout ________________________________________________________________________
> 
> device = <tests.conftest.TestDevice object at 0x7f75a2025610>, resetDevice = <tests.conftest.TestDevice object at 0x7f75a2025610>
> 
>     @pytest.mark.skipif(
>         "trezor" in sys.argv, reason="ClientPin is not supported on Trezor."
>     )
>     def test_lockout(device, resetDevice):
>         pin = "TestPin"
>         device.client.pin_protocol.set_pin(pin)
>     
>         pin_token = device.client.pin_protocol.get_pin_token(pin)
>         req = FidoRequest(pin_token=pin_token)
>     
>         req.pin_auth = hmac_sha256(pin_token, req.cdh)[:16]
>     
>         for i in range(1, 10):
>             err = CtapError.ERR.PIN_INVALID
>             if i in (3, 6):
>                 err = CtapError.ERR.PIN_AUTH_BLOCKED
>             elif i >= 8:
>                 err = [CtapError.ERR.PIN_BLOCKED, CtapError.ERR.PIN_INVALID]
>     
>             with pytest.raises(CtapError) as e:
>                 device.sendPP("WrongPin")
>             assert e.value.code == err or e.value.code in err
>     
>             attempts = 8 - i
>             if i > 8:
>                 attempts = 0
>     
>             res = device.ctap2.client_pin(1, PinProtocolV1.CMD.GET_RETRIES)
>             assert res[3] == attempts
>     
>             if err == CtapError.ERR.PIN_AUTH_BLOCKED:
> >               device.reboot()
> 
> tests/standard/fido2/pin/test_lockout.py:40: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> tests/conftest.py:302: in reboot
>     self.find_device(self.nfc_interface_only)
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> 
> self = <tests.conftest.TestDevice object at 0x7f75a2025610>, nfcInterfaceOnly = False
> 
>     def find_device(self, nfcInterfaceOnly=False):
>         dev = None
>         self.nfc_interface_only = nfcInterfaceOnly
>         if not nfcInterfaceOnly:
>             print("--- HID ---")
>             print(list(CtapHidDevice.list_devices()))
>             dev = next(CtapHidDevice.list_devices(), None)
>     
>         '''
>         else:
>             from fido2.pcsc import CtapPcscDevice
>     
>             print("--- NFC ---")
>             dev = next(MoreRobustPcscDevice.list_devices(), None)
>     
>             if dev:
>                 self.is_nfc = True
>                 # For ACR1252 readers, with drivers installed
>                 # https://www.acs.com.hk/en/products/342/acr1252u-usb-nfc-reader-iii-nfc-forum-certified-reader
>                 # disable auto pps, always use 106kbps
>                 # dev.control_exchange(SCARD_CTL_CODE(0x3500), b"\xE0\x00\x00\x24\x02\x00\x00")
>         '''
>     
>         if not dev:
> >           raise RuntimeError("No FIDO device found")
> E           RuntimeError: No FIDO device found
> 
> tests/conftest.py:259: RuntimeError
> __________________________________________________________________ TestSetPin.test_set_pin __________________________________________________________________
> 
> self = <tests.standard.fido2.pin.test_set_pin.TestSetPin object at 0x7f75a14d40a0>, device = <tests.conftest.TestDevice object at 0x7f75a2025610>
> 
>     def test_set_pin(self, device):
> >       device.client.pin_protocol.set_pin("TestPin")
> 
> tests/standard/fido2/pin/test_set_pin.py:23: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> venv/lib/python3.8/site-packages/fido2/ctap2.py:940: in set_pin
>     key_agreement, shared_secret = self.get_shared_secret()
> venv/lib/python3.8/site-packages/fido2/ctap2.py:885: in get_shared_secret
>     resp = self.ctap.client_pin(
> venv/lib/python3.8/site-packages/fido2/ctap2.py:776: in client_pin
>     return self.send_cbor(
> venv/lib/python3.8/site-packages/fido2/ctap2.py:642: in send_cbor
>     response = self.device.call(CTAPHID.CBOR, request, event, on_keepalive)
> venv/lib/python3.8/site-packages/fido2/hid.py:86: in call
>     self._dev.InternalSend(TYPE_INIT | cmd, bytearray(data))
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:277: in InternalSend
>     self.InternalSendPacket(first_packet)
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:293: in InternalSendPacket
>     self.hid_device.Write(wire)
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> 
> self = <fido2._pyu2f.linux.LinuxHidDevice object at 0x7f75a155f250>, packet = [2, 0, 0, 0, 144, 0, ...]
> 
>     def Write(self, packet):
>       """See base class."""
>       out = bytes(bytearray([0] + packet))  # Prepend the zero-byte (report ID)
> >     os.write(self.dev, out)
> E     OSError: [Errno 19] No such device
> 
> venv/lib/python3.8/site-packages/fido2/_pyu2f/linux.py:222: OSError
> ______________________________________________________________ TestSetPin.test_set_pin_too_big ______________________________________________________________
> 
> self = <tests.standard.fido2.pin.test_set_pin.TestSetPin object at 0x7f75a13c0490>, device = <tests.conftest.TestDevice object at 0x7f75a2025610>
> 
>     def test_set_pin_too_big(self, device):
>         with pytest.raises(CtapError) as e:
> >           device.client.pin_protocol.set_pin("A" * 64)
> 
> tests/standard/fido2/pin/test_set_pin.py:28: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> venv/lib/python3.8/site-packages/fido2/ctap2.py:940: in set_pin
>     key_agreement, shared_secret = self.get_shared_secret()
> venv/lib/python3.8/site-packages/fido2/ctap2.py:885: in get_shared_secret
>     resp = self.ctap.client_pin(
> venv/lib/python3.8/site-packages/fido2/ctap2.py:776: in client_pin
>     return self.send_cbor(
> venv/lib/python3.8/site-packages/fido2/ctap2.py:642: in send_cbor
>     response = self.device.call(CTAPHID.CBOR, request, event, on_keepalive)
> venv/lib/python3.8/site-packages/fido2/hid.py:86: in call
>     self._dev.InternalSend(TYPE_INIT | cmd, bytearray(data))
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:277: in InternalSend
>     self.InternalSendPacket(first_packet)
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:293: in InternalSendPacket
>     self.hid_device.Write(wire)
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> 
> self = <fido2._pyu2f.linux.LinuxHidDevice object at 0x7f75a155f250>, packet = [2, 0, 0, 0, 144, 0, ...]
> 
>     def Write(self, packet):
>       """See base class."""
>       out = bytes(bytearray([0] + packet))  # Prepend the zero-byte (report ID)
> >     os.write(self.dev, out)
> E     OSError: [Errno 19] No such device
> 
> venv/lib/python3.8/site-packages/fido2/_pyu2f/linux.py:222: OSError
> _______________________________________________________ TestSetPin.test_get_pin_token_but_no_pin_set ________________________________________________________
> 
> self = <tests.standard.fido2.pin.test_set_pin.TestSetPin object at 0x7f75a12d7070>, device = <tests.conftest.TestDevice object at 0x7f75a2025610>
> 
>     def test_get_pin_token_but_no_pin_set(self, device):
>         with pytest.raises(CtapError) as e:
> >           device.client.pin_protocol.get_pin_token("TestPin")
> 
> tests/standard/fido2/pin/test_set_pin.py:33: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> venv/lib/python3.8/site-packages/fido2/ctap2.py:905: in get_pin_token
>     key_agreement, shared_secret = self.get_shared_secret()
> venv/lib/python3.8/site-packages/fido2/ctap2.py:885: in get_shared_secret
>     resp = self.ctap.client_pin(
> venv/lib/python3.8/site-packages/fido2/ctap2.py:776: in client_pin
>     return self.send_cbor(
> venv/lib/python3.8/site-packages/fido2/ctap2.py:642: in send_cbor
>     response = self.device.call(CTAPHID.CBOR, request, event, on_keepalive)
> venv/lib/python3.8/site-packages/fido2/hid.py:86: in call
>     self._dev.InternalSend(TYPE_INIT | cmd, bytearray(data))
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:277: in InternalSend
>     self.InternalSendPacket(first_packet)
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:293: in InternalSendPacket
>     self.hid_device.Write(wire)
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> 
> self = <fido2._pyu2f.linux.LinuxHidDevice object at 0x7f75a155f250>, packet = [2, 0, 0, 0, 144, 0, ...]
> 
>     def Write(self, packet):
>       """See base class."""
>       out = bytes(bytearray([0] + packet))  # Prepend the zero-byte (report ID)
> >     os.write(self.dev, out)
> E     OSError: [Errno 19] No such device
> 
> venv/lib/python3.8/site-packages/fido2/_pyu2f/linux.py:222: OSError
> _________________________________________________________ TestSetPin.test_change_pin_but_no_pin_set _________________________________________________________
> 
> self = <tests.standard.fido2.pin.test_set_pin.TestSetPin object at 0x7f75a1371eb0>, device = <tests.conftest.TestDevice object at 0x7f75a2025610>
> 
>     def test_change_pin_but_no_pin_set(self, device):
>         with pytest.raises(CtapError) as e:
> >           device.client.pin_protocol.change_pin("TestPin", "1234")
> 
> tests/standard/fido2/pin/test_set_pin.py:38: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> venv/lib/python3.8/site-packages/fido2/ctap2.py:964: in change_pin
>     key_agreement, shared_secret = self.get_shared_secret()
> venv/lib/python3.8/site-packages/fido2/ctap2.py:885: in get_shared_secret
>     resp = self.ctap.client_pin(
> venv/lib/python3.8/site-packages/fido2/ctap2.py:776: in client_pin
>     return self.send_cbor(
> venv/lib/python3.8/site-packages/fido2/ctap2.py:642: in send_cbor
>     response = self.device.call(CTAPHID.CBOR, request, event, on_keepalive)
> venv/lib/python3.8/site-packages/fido2/hid.py:86: in call
>     self._dev.InternalSend(TYPE_INIT | cmd, bytearray(data))
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:277: in InternalSend
>     self.InternalSendPacket(first_packet)
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:293: in InternalSendPacket
>     self.hid_device.Write(wire)
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> 
> self = <fido2._pyu2f.linux.LinuxHidDevice object at 0x7f75a155f250>, packet = [2, 0, 0, 0, 144, 0, ...]
> 
>     def Write(self, packet):
>       """See base class."""
>       out = bytes(bytearray([0] + packet))  # Prepend the zero-byte (report ID)
> >     os.write(self.dev, out)
> E     OSError: [Errno 19] No such device
> 
> venv/lib/python3.8/site-packages/fido2/_pyu2f/linux.py:222: OSError
> _________________________________________________________ TestSetPin.test_setting_pin_and_get_info __________________________________________________________
> 
> self = <tests.standard.fido2.pin.test_set_pin.TestSetPin object at 0x7f75a153abe0>, device = <tests.conftest.TestDevice object at 0x7f75a2025610>
> 
>     def test_setting_pin_and_get_info(self, device):
> >       device.client.pin_protocol.set_pin("TestPin")
> 
> tests/standard/fido2/pin/test_set_pin.py:42: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> venv/lib/python3.8/site-packages/fido2/ctap2.py:940: in set_pin
>     key_agreement, shared_secret = self.get_shared_secret()
> venv/lib/python3.8/site-packages/fido2/ctap2.py:885: in get_shared_secret
>     resp = self.ctap.client_pin(
> venv/lib/python3.8/site-packages/fido2/ctap2.py:776: in client_pin
>     return self.send_cbor(
> venv/lib/python3.8/site-packages/fido2/ctap2.py:642: in send_cbor
>     response = self.device.call(CTAPHID.CBOR, request, event, on_keepalive)
> venv/lib/python3.8/site-packages/fido2/hid.py:86: in call
>     self._dev.InternalSend(TYPE_INIT | cmd, bytearray(data))
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:277: in InternalSend
>     self.InternalSendPacket(first_packet)
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:293: in InternalSendPacket
>     self.hid_device.Write(wire)
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> 
> self = <fido2._pyu2f.linux.LinuxHidDevice object at 0x7f75a155f250>, packet = [2, 0, 0, 0, 144, 0, ...]
> 
>     def Write(self, packet):
>       """See base class."""
>       out = bytes(bytearray([0] + packet))  # Prepend the zero-byte (report ID)
> >     os.write(self.dev, out)
> E     OSError: [Errno 19] No such device
> 
> venv/lib/python3.8/site-packages/fido2/_pyu2f/linux.py:222: OSError
> __________________________________________________________________ TestHID.test_long_ping ___________________________________________________________________
> 
> self = <tests.standard.transport.test_hid.TestHID object at 0x7f75a13e27c0>, device = <tests.conftest.TestDevice object at 0x7f75a2025610>
> 
>     def test_long_ping(self, device):
>         amt = 1000
>         pingdata = os.urandom(amt)
>     
>         t1 = time.time() * 1000
> >       r = device.send_data(CTAPHID.PING, pingdata)
> 
> tests/standard/transport/test_hid.py:19: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> tests/conftest.py:309: in send_data
>     return self.dev.call(cmd, data, event, on_keepalive = on_keepalive)
> venv/lib/python3.8/site-packages/fido2/hid.py:86: in call
>     self._dev.InternalSend(TYPE_INIT | cmd, bytearray(data))
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:277: in InternalSend
>     self.InternalSendPacket(first_packet)
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:293: in InternalSendPacket
>     self.hid_device.Write(wire)
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> 
> self = <fido2._pyu2f.linux.LinuxHidDevice object at 0x7f75a155f250>, packet = [2, 0, 0, 0, 129, 3, ...]
> 
>     def Write(self, packet):
>       """See base class."""
>       out = bytes(bytearray([0] + packet))  # Prepend the zero-byte (report ID)
> >     os.write(self.dev, out)
> E     OSError: [Errno 19] No such device
> 
> venv/lib/python3.8/site-packages/fido2/_pyu2f/linux.py:222: OSError
> _____________________________________________________________________ TestHID.test_init _____________________________________________________________________
> 
> self = <tests.standard.transport.test_hid.TestHID object at 0x7f75a13c5340>, device = <tests.conftest.TestDevice object at 0x7f75a2025610>
> check_timeouts = False
> 
>     def test_init(self, device, check_timeouts=False):
>         if check_timeouts:
>             with pytest.raises(socket.timeout):
>                 cmd, resp = self.recv_raw()
>     
>         payload = b"\x11\x11\x11\x11\x11\x11\x11\x11"
> >       r = device.send_data(CTAPHID.INIT, payload)
> 
> tests/standard/transport/test_hid.py:33: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> tests/conftest.py:309: in send_data
>     return self.dev.call(cmd, data, event, on_keepalive = on_keepalive)
> venv/lib/python3.8/site-packages/fido2/hid.py:86: in call
>     self._dev.InternalSend(TYPE_INIT | cmd, bytearray(data))
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:277: in InternalSend
>     self.InternalSendPacket(first_packet)
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:293: in InternalSendPacket
>     self.hid_device.Write(wire)
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> 
> self = <fido2._pyu2f.linux.LinuxHidDevice object at 0x7f75a155f250>, packet = [2, 0, 0, 0, 134, 0, ...]
> 
>     def Write(self, packet):
>       """See base class."""
>       out = bytes(bytearray([0] + packet))  # Prepend the zero-byte (report ID)
> >     os.write(self.dev, out)
> E     OSError: [Errno 19] No such device
> 
> venv/lib/python3.8/site-packages/fido2/_pyu2f/linux.py:222: OSError
> _____________________________________________________________________ TestHID.test_ping _____________________________________________________________________
> 
> self = <tests.standard.transport.test_hid.TestHID object at 0x7f75a128b670>, device = <tests.conftest.TestDevice object at 0x7f75a2025610>
> 
>     def test_ping(self, device):
>     
>         pingdata = os.urandom(100)
> >       r = device.send_data(CTAPHID.PING, pingdata)
> 
> tests/standard/transport/test_hid.py:40: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> tests/conftest.py:309: in send_data
>     return self.dev.call(cmd, data, event, on_keepalive = on_keepalive)
> venv/lib/python3.8/site-packages/fido2/hid.py:86: in call
>     self._dev.InternalSend(TYPE_INIT | cmd, bytearray(data))
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:277: in InternalSend
>     self.InternalSendPacket(first_packet)
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:293: in InternalSendPacket
>     self.hid_device.Write(wire)
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> 
> self = <fido2._pyu2f.linux.LinuxHidDevice object at 0x7f75a155f250>, packet = [2, 0, 0, 0, 129, 0, ...]
> 
>     def Write(self, packet):
>       """See base class."""
>       out = bytes(bytearray([0] + packet))  # Prepend the zero-byte (report ID)
> >     os.write(self.dev, out)
> E     OSError: [Errno 19] No such device
> 
> venv/lib/python3.8/site-packages/fido2/_pyu2f/linux.py:222: OSError
> _____________________________________________________________________ TestHID.test_wink _____________________________________________________________________
> 
> self = <tests.standard.transport.test_hid.TestHID object at 0x7f75a1457c40>, device = <tests.conftest.TestDevice object at 0x7f75a2025610>
> 
>     def test_wink(self, device):
> >       r = device.send_data(CTAPHID.WINK, "")
> 
> tests/standard/transport/test_hid.py:44: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> tests/conftest.py:309: in send_data
>     return self.dev.call(cmd, data, event, on_keepalive = on_keepalive)
> venv/lib/python3.8/site-packages/fido2/hid.py:86: in call
>     self._dev.InternalSend(TYPE_INIT | cmd, bytearray(data))
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:277: in InternalSend
>     self.InternalSendPacket(first_packet)
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:293: in InternalSendPacket
>     self.hid_device.Write(wire)
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> 
> self = <fido2._pyu2f.linux.LinuxHidDevice object at 0x7f75a155f250>, packet = [2, 0, 0, 0, 136, 0, ...]
> 
>     def Write(self, packet):
>       """See base class."""
>       out = bytes(bytearray([0] + packet))  # Prepend the zero-byte (report ID)
> >     os.write(self.dev, out)
> E     OSError: [Errno 19] No such device
> 
> venv/lib/python3.8/site-packages/fido2/_pyu2f/linux.py:222: OSError
> _______________________________________________________________ TestHID.test_cbor_no_payload ________________________________________________________________
> 
> self = <tests.standard.transport.test_hid.TestHID object at 0x7f75a1471130>, device = <tests.conftest.TestDevice object at 0x7f75a2025610>
> 
>     def test_cbor_no_payload(self, device):
>         payload = b"\x11\x11\x11\x11\x11\x11\x11\x11"
> >       r = device.send_data(CTAPHID.INIT, payload)
> 
> tests/standard/transport/test_hid.py:48: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> tests/conftest.py:309: in send_data
>     return self.dev.call(cmd, data, event, on_keepalive = on_keepalive)
> venv/lib/python3.8/site-packages/fido2/hid.py:86: in call
>     self._dev.InternalSend(TYPE_INIT | cmd, bytearray(data))
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:277: in InternalSend
>     self.InternalSendPacket(first_packet)
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:293: in InternalSendPacket
>     self.hid_device.Write(wire)
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> 
> self = <fido2._pyu2f.linux.LinuxHidDevice object at 0x7f75a155f250>, packet = [2, 0, 0, 0, 134, 0, ...]
> 
>     def Write(self, packet):
>       """See base class."""
>       out = bytes(bytearray([0] + packet))  # Prepend the zero-byte (report ID)
> >     os.write(self.dev, out)
> E     OSError: [Errno 19] No such device
> 
> venv/lib/python3.8/site-packages/fido2/_pyu2f/linux.py:222: OSError
> ______________________________________________________________ TestHID.test_no_data_in_u2f_msg ______________________________________________________________
> 
> self = <tests.standard.transport.test_hid.TestHID object at 0x7f75a1433100>, device = <tests.conftest.TestDevice object at 0x7f75a2025610>
> 
>     def test_no_data_in_u2f_msg(self, device):
>         payload = b"\x11\x11\x11\x11\x11\x11\x11\x11"
> >       r = device.send_data(CTAPHID.INIT, payload)
> 
> tests/standard/transport/test_hid.py:61: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> tests/conftest.py:309: in send_data
>     return self.dev.call(cmd, data, event, on_keepalive = on_keepalive)
> venv/lib/python3.8/site-packages/fido2/hid.py:86: in call
>     self._dev.InternalSend(TYPE_INIT | cmd, bytearray(data))
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:277: in InternalSend
>     self.InternalSendPacket(first_packet)
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:293: in InternalSendPacket
>     self.hid_device.Write(wire)
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> 
> self = <fido2._pyu2f.linux.LinuxHidDevice object at 0x7f75a155f250>, packet = [2, 0, 0, 0, 134, 0, ...]
> 
>     def Write(self, packet):
>       """See base class."""
>       out = bytes(bytearray([0] + packet))  # Prepend the zero-byte (report ID)
> >     os.write(self.dev, out)
> E     OSError: [Errno 19] No such device
> 
> venv/lib/python3.8/site-packages/fido2/_pyu2f/linux.py:222: OSError
> _______________________________________________________________ TestHID.test_invalid_hid_cmd ________________________________________________________________
> 
> self = <tests.standard.transport.test_hid.TestHID object at 0x7f75a131f220>, device = <tests.conftest.TestDevice object at 0x7f75a2025610>
> 
>     def test_invalid_hid_cmd(self, device):
> >       r = device.send_data(CTAPHID.INIT, "\x11\x22\x33\x44\x55\x66\x77\x88")
> 
> tests/standard/transport/test_hid.py:74: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> tests/conftest.py:309: in send_data
>     return self.dev.call(cmd, data, event, on_keepalive = on_keepalive)
> venv/lib/python3.8/site-packages/fido2/hid.py:86: in call
>     self._dev.InternalSend(TYPE_INIT | cmd, bytearray(data))
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:277: in InternalSend
>     self.InternalSendPacket(first_packet)
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:293: in InternalSendPacket
>     self.hid_device.Write(wire)
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> 
> self = <fido2._pyu2f.linux.LinuxHidDevice object at 0x7f75a155f250>, packet = [2, 0, 0, 0, 134, 0, ...]
> 
>     def Write(self, packet):
>       """See base class."""
>       out = bytes(bytearray([0] + packet))  # Prepend the zero-byte (report ID)
> >     os.write(self.dev, out)
> E     OSError: [Errno 19] No such device
> 
> venv/lib/python3.8/site-packages/fido2/_pyu2f/linux.py:222: OSError
> _______________________________________________________________ TestHID.test_oversize_packet ________________________________________________________________
> 
> self = <tests.standard.transport.test_hid.TestHID object at 0x7f75a1457b80>, device = <tests.conftest.TestDevice object at 0x7f75a2025610>
> 
>     def test_oversize_packet(self, device):
> >       device.send_raw("\x81\x1d\xba\x00")
> 
> tests/standard/transport/test_hid.py:81: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> tests/conftest.py:326: in send_raw
>     self.dev._dev.InternalSendPacket(Packet(data))
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:293: in InternalSendPacket
>     self.hid_device.Write(wire)
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> 
> self = <fido2._pyu2f.linux.LinuxHidDevice object at 0x7f75a155f250>, packet = [2, 0, 0, 0, 129, 29, ...]
> 
>     def Write(self, packet):
>       """See base class."""
>       out = bytes(bytearray([0] + packet))  # Prepend the zero-byte (report ID)
> >     os.write(self.dev, out)
> E     OSError: [Errno 19] No such device
> 
> venv/lib/python3.8/site-packages/fido2/_pyu2f/linux.py:222: OSError
> _____________________________________________________________ TestHID.test_skip_sequence_number _____________________________________________________________
> 
> self = <tests.standard.transport.test_hid.TestHID object at 0x7f75a14c6610>, device = <tests.conftest.TestDevice object at 0x7f75a2025610>
> 
>     def test_skip_sequence_number(self, device):
> >       r = device.send_data(CTAPHID.PING, "\x44" * 200)
> 
> tests/standard/transport/test_hid.py:86: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> tests/conftest.py:309: in send_data
>     return self.dev.call(cmd, data, event, on_keepalive = on_keepalive)
> venv/lib/python3.8/site-packages/fido2/hid.py:86: in call
>     self._dev.InternalSend(TYPE_INIT | cmd, bytearray(data))
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:277: in InternalSend
>     self.InternalSendPacket(first_packet)
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:293: in InternalSendPacket
>     self.hid_device.Write(wire)
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> 
> self = <fido2._pyu2f.linux.LinuxHidDevice object at 0x7f75a155f250>, packet = [2, 0, 0, 0, 129, 0, ...]
> 
>     def Write(self, packet):
>       """See base class."""
>       out = bytes(bytearray([0] + packet))  # Prepend the zero-byte (report ID)
> >     os.write(self.dev, out)
> E     OSError: [Errno 19] No such device
> 
> venv/lib/python3.8/site-packages/fido2/_pyu2f/linux.py:222: OSError
> _______________________________________________________________ TestHID.test_resync_and_ping ________________________________________________________________
> 
> self = <tests.standard.transport.test_hid.TestHID object at 0x7f75a12a7c70>, device = <tests.conftest.TestDevice object at 0x7f75a2025610>
> 
>     def test_resync_and_ping(self, device):
> >       r = device.send_data(CTAPHID.INIT, "\x11\x22\x33\x44\x55\x66\x77\x88")
> 
> tests/standard/transport/test_hid.py:96: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> tests/conftest.py:309: in send_data
>     return self.dev.call(cmd, data, event, on_keepalive = on_keepalive)
> venv/lib/python3.8/site-packages/fido2/hid.py:86: in call
>     self._dev.InternalSend(TYPE_INIT | cmd, bytearray(data))
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:277: in InternalSend
>     self.InternalSendPacket(first_packet)
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:293: in InternalSendPacket
>     self.hid_device.Write(wire)
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> 
> self = <fido2._pyu2f.linux.LinuxHidDevice object at 0x7f75a155f250>, packet = [2, 0, 0, 0, 134, 0, ...]
> 
>     def Write(self, packet):
>       """See base class."""
>       out = bytes(bytearray([0] + packet))  # Prepend the zero-byte (report ID)
> >     os.write(self.dev, out)
> E     OSError: [Errno 19] No such device
> 
> venv/lib/python3.8/site-packages/fido2/_pyu2f/linux.py:222: OSError
> __________________________________________________________________ TestHID.test_ping_abort __________________________________________________________________
> 
> self = <tests.standard.transport.test_hid.TestHID object at 0x7f75a12d7cd0>, device = <tests.conftest.TestDevice object at 0x7f75a2025610>
> 
>     def test_ping_abort(self, device):
> >       device.send_raw("\x81\x04\x00")
> 
> tests/standard/transport/test_hid.py:103: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> tests/conftest.py:326: in send_raw
>     self.dev._dev.InternalSendPacket(Packet(data))
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:293: in InternalSendPacket
>     self.hid_device.Write(wire)
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> 
> self = <fido2._pyu2f.linux.LinuxHidDevice object at 0x7f75a155f250>, packet = [2, 0, 0, 0, 129, 4, ...]
> 
>     def Write(self, packet):
>       """See base class."""
>       out = bytes(bytearray([0] + packet))  # Prepend the zero-byte (report ID)
> >     os.write(self.dev, out)
> E     OSError: [Errno 19] No such device
> 
> venv/lib/python3.8/site-packages/fido2/_pyu2f/linux.py:222: OSError
> ___________________________________________________________________ TestHID.test_timeout ____________________________________________________________________
> 
> self = <tests.standard.transport.test_hid.TestHID object at 0x7f75a131f1c0>, device = <tests.conftest.TestDevice object at 0x7f75a2025610>
> 
>     def test_timeout(self, device):
> >       device.send_data(CTAPHID.INIT, "\x11\x22\x33\x44\x55\x66\x77\x88")
> 
> tests/standard/transport/test_hid.py:139: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> tests/conftest.py:309: in send_data
>     return self.dev.call(cmd, data, event, on_keepalive = on_keepalive)
> venv/lib/python3.8/site-packages/fido2/hid.py:86: in call
>     self._dev.InternalSend(TYPE_INIT | cmd, bytearray(data))
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:277: in InternalSend
>     self.InternalSendPacket(first_packet)
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:293: in InternalSendPacket
>     self.hid_device.Write(wire)
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> 
> self = <fido2._pyu2f.linux.LinuxHidDevice object at 0x7f75a155f250>, packet = [2, 0, 0, 0, 134, 0, ...]
> 
>     def Write(self, packet):
>       """See base class."""
>       out = bytes(bytearray([0] + packet))  # Prepend the zero-byte (report ID)
> >     os.write(self.dev, out)
> E     OSError: [Errno 19] No such device
> 
> venv/lib/python3.8/site-packages/fido2/_pyu2f/linux.py:222: OSError
> ___________________________________________________________________ TestHID.test_not_cont ___________________________________________________________________
> 
> self = <tests.standard.transport.test_hid.TestHID object at 0x7f75a12b31f0>, device = <tests.conftest.TestDevice object at 0x7f75a2025610>
> check_timeouts = False
> 
>     def test_not_cont(self, device, check_timeouts=False):
> >       device.send_data(CTAPHID.INIT, "\x11\x22\x33\x44\x55\x66\x77\x88")
> 
> tests/standard/transport/test_hid.py:152: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> tests/conftest.py:309: in send_data
>     return self.dev.call(cmd, data, event, on_keepalive = on_keepalive)
> venv/lib/python3.8/site-packages/fido2/hid.py:86: in call
>     self._dev.InternalSend(TYPE_INIT | cmd, bytearray(data))
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:277: in InternalSend
>     self.InternalSendPacket(first_packet)
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:293: in InternalSendPacket
>     self.hid_device.Write(wire)
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> 
> self = <fido2._pyu2f.linux.LinuxHidDevice object at 0x7f75a155f250>, packet = [2, 0, 0, 0, 134, 0, ...]
> 
>     def Write(self, packet):
>       """See base class."""
>       out = bytes(bytearray([0] + packet))  # Prepend the zero-byte (report ID)
> >     os.write(self.dev, out)
> E     OSError: [Errno 19] No such device
> 
> venv/lib/python3.8/site-packages/fido2/_pyu2f/linux.py:222: OSError
> __________________________________________________________________ TestHID.test_check_busy __________________________________________________________________
> 
> self = <tests.standard.transport.test_hid.TestHID object at 0x7f75a13925b0>, device = <tests.conftest.TestDevice object at 0x7f75a2025610>
> 
>     def test_check_busy(self, device):
>         t1 = time.time() * 1000
> >       device.send_data(CTAPHID.INIT, "\x11\x22\x33\x44\x55\x66\x77\x88")
> 
> tests/standard/transport/test_hid.py:169: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> tests/conftest.py:309: in send_data
>     return self.dev.call(cmd, data, event, on_keepalive = on_keepalive)
> venv/lib/python3.8/site-packages/fido2/hid.py:86: in call
>     self._dev.InternalSend(TYPE_INIT | cmd, bytearray(data))
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:277: in InternalSend
>     self.InternalSendPacket(first_packet)
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:293: in InternalSendPacket
>     self.hid_device.Write(wire)
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> 
> self = <fido2._pyu2f.linux.LinuxHidDevice object at 0x7f75a155f250>, packet = [2, 0, 0, 0, 134, 0, ...]
> 
>     def Write(self, packet):
>       """See base class."""
>       out = bytes(bytearray([0] + packet))  # Prepend the zero-byte (report ID)
> >     os.write(self.dev, out)
> E     OSError: [Errno 19] No such device
> 
> venv/lib/python3.8/site-packages/fido2/_pyu2f/linux.py:222: OSError
> ____________________________________________________________ TestHID.test_check_busy_interleaved ____________________________________________________________
> 
> self = <tests.standard.transport.test_hid.TestHID object at 0x7f75a12aecd0>, device = <tests.conftest.TestDevice object at 0x7f75a2025610>
> 
>     def test_check_busy_interleaved(self, device):
>         cid1 = "\x11\x22\x33\x44"
>         cid2 = "\x01\x22\x33\x44"
>         device.set_cid(cid2)
> >       device.send_data(CTAPHID.INIT, "\x11\x22\x33\x44\x55\x66\x77\x88")
> 
> tests/standard/transport/test_hid.py:190: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> tests/conftest.py:309: in send_data
>     return self.dev.call(cmd, data, event, on_keepalive = on_keepalive)
> venv/lib/python3.8/site-packages/fido2/hid.py:86: in call
>     self._dev.InternalSend(TYPE_INIT | cmd, bytearray(data))
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:277: in InternalSend
>     self.InternalSendPacket(first_packet)
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:293: in InternalSendPacket
>     self.hid_device.Write(wire)
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> 
> self = <fido2._pyu2f.linux.LinuxHidDevice object at 0x7f75a155f250>, packet = [1, 34, 51, 68, 134, 0, ...]
> 
>     def Write(self, packet):
>       """See base class."""
>       out = bytes(bytearray([0] + packet))  # Prepend the zero-byte (report ID)
> >     os.write(self.dev, out)
> E     OSError: [Errno 19] No such device
> 
> venv/lib/python3.8/site-packages/fido2/_pyu2f/linux.py:222: OSError
> ____________________________________________________________________ TestHID.test_cid_0 _____________________________________________________________________
> 
> self = <tests.standard.transport.test_hid.TestHID object at 0x7f75a128b1c0>, device = <tests.conftest.TestDevice object at 0x7f75a2025610>
> 
>     def test_cid_0(self, device):
>         device.set_cid("\x00\x00\x00\x00")
> >       device.send_raw(
>             "\x86\x00\x08\x11\x22\x33\x44\x55\x66\x77\x88", cid="\x00\x00\x00\x00"
>         )
> 
> tests/standard/transport/test_hid.py:217: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> tests/conftest.py:326: in send_raw
>     self.dev._dev.InternalSendPacket(Packet(data))
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:293: in InternalSendPacket
>     self.hid_device.Write(wire)
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> 
> self = <fido2._pyu2f.linux.LinuxHidDevice object at 0x7f75a155f250>, packet = [0, 0, 0, 0, 134, 0, ...]
> 
>     def Write(self, packet):
>       """See base class."""
>       out = bytes(bytearray([0] + packet))  # Prepend the zero-byte (report ID)
> >     os.write(self.dev, out)
> E     OSError: [Errno 19] No such device
> 
> venv/lib/python3.8/site-packages/fido2/_pyu2f/linux.py:222: OSError
> _________________________________________________________________ TestHID.test_cid_ffffffff _________________________________________________________________
> 
> self = <tests.standard.transport.test_hid.TestHID object at 0x7f75a2d9f580>, device = <tests.conftest.TestDevice object at 0x7f75a2025610>
> 
>     def test_cid_ffffffff(self, device):
>     
>         device.set_cid("\xff\xff\xff\xff")
> >       device.send_raw(
>             "\x81\x00\x08\x11\x22\x33\x44\x55\x66\x77\x88", cid="\xff\xff\xff\xff"
>         )
> 
> tests/standard/transport/test_hid.py:228: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> tests/conftest.py:326: in send_raw
>     self.dev._dev.InternalSendPacket(Packet(data))
> venv/lib/python3.8/site-packages/fido2/_pyu2f/hidtransport.py:293: in InternalSendPacket
>     self.hid_device.Write(wire)
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> 
> self = <fido2._pyu2f.linux.LinuxHidDevice object at 0x7f75a155f250>, packet = [255, 255, 255, 255, 129, 0, ...]
> 
>     def Write(self, packet):
>       """See base class."""
>       out = bytes(bytearray([0] + packet))  # Prepend the zero-byte (report ID)
> >     os.write(self.dev, out)
> E     OSError: [Errno 19] No such device
> 
> venv/lib/python3.8/site-packages/fido2/_pyu2f/linux.py:222: OSError
> ================================================================== short test summary info ==================================================================
> FAILED tests/standard/fido2/pin/test_lockout.py::test_lockout - RuntimeError: No FIDO device found
> FAILED tests/standard/fido2/pin/test_set_pin.py::TestSetPin::test_set_pin - OSError: [Errno 19] No such device
> FAILED tests/standard/fido2/pin/test_set_pin.py::TestSetPin::test_set_pin_too_big - OSError: [Errno 19] No such device
> FAILED tests/standard/fido2/pin/test_set_pin.py::TestSetPin::test_get_pin_token_but_no_pin_set - OSError: [Errno 19] No such device
> FAILED tests/standard/fido2/pin/test_set_pin.py::TestSetPin::test_change_pin_but_no_pin_set - OSError: [Errno 19] No such device
> FAILED tests/standard/fido2/pin/test_set_pin.py::TestSetPin::test_setting_pin_and_get_info - OSError: [Errno 19] No such device
> FAILED tests/standard/transport/test_hid.py::TestHID::test_long_ping - OSError: [Errno 19] No such device
> FAILED tests/standard/transport/test_hid.py::TestHID::test_init - OSError: [Errno 19] No such device
> FAILED tests/standard/transport/test_hid.py::TestHID::test_ping - OSError: [Errno 19] No such device
> FAILED tests/standard/transport/test_hid.py::TestHID::test_wink - OSError: [Errno 19] No such device
> FAILED tests/standard/transport/test_hid.py::TestHID::test_cbor_no_payload - OSError: [Errno 19] No such device
> FAILED tests/standard/transport/test_hid.py::TestHID::test_no_data_in_u2f_msg - OSError: [Errno 19] No such device
> FAILED tests/standard/transport/test_hid.py::TestHID::test_invalid_hid_cmd - OSError: [Errno 19] No such device
> FAILED tests/standard/transport/test_hid.py::TestHID::test_oversize_packet - OSError: [Errno 19] No such device
> FAILED tests/standard/transport/test_hid.py::TestHID::test_skip_sequence_number - OSError: [Errno 19] No such device
> FAILED tests/standard/transport/test_hid.py::TestHID::test_resync_and_ping - OSError: [Errno 19] No such device
> FAILED tests/standard/transport/test_hid.py::TestHID::test_ping_abort - OSError: [Errno 19] No such device
> FAILED tests/standard/transport/test_hid.py::TestHID::test_timeout - OSError: [Errno 19] No such device
> FAILED tests/standard/transport/test_hid.py::TestHID::test_not_cont - OSError: [Errno 19] No such device
> FAILED tests/standard/transport/test_hid.py::TestHID::test_check_busy - OSError: [Errno 19] No such device
> FAILED tests/standard/transport/test_hid.py::TestHID::test_check_busy_interleaved - OSError: [Errno 19] No such device
> FAILED tests/standard/transport/test_hid.py::TestHID::test_cid_0 - OSError: [Errno 19] No such device
> FAILED tests/standard/transport/test_hid.py::TestHID::test_cid_ffffffff - OSError: [Errno 19] No such device
> ERROR tests/standard/fido2/pin/test_pin.py::TestPin::test_pin[123456789A] - OSError: [Errno 19] No such device
> ERROR tests/standard/fido2/pin/test_pin.py::TestPin::test_set_pin_twice[123456789A] - OSError: [Errno 19] No such device
> ERROR tests/standard/fido2/pin/test_pin.py::TestPin::test_get_key_agreement_fields[123456789A] - OSError: [Errno 19] No such device
> ERROR tests/standard/fido2/pin/test_pin.py::TestPin::test_verify_flag[123456789A] - OSError: [Errno 19] No such device
> ERROR tests/standard/fido2/pin/test_pin.py::TestPin::test_get_no_pin_auth[123456789A] - OSError: [Errno 19] No such device
> ERROR tests/standard/fido2/pin/test_pin.py::TestPin::test_zero_length_pin_auth[123456789A] - OSError: [Errno 19] No such device
> ERROR tests/standard/fido2/pin/test_pin.py::TestPin::test_make_credential_no_pin[123456789A] - OSError: [Errno 19] No such device
> ERROR tests/standard/fido2/pin/test_pin.py::TestPin::test_get_assertion_no_pin[123456789A] - OSError: [Errno 19] No such device
> ERROR tests/standard/fido2/pin/test_pin.py::TestChangePin::test_change_pin[123456789A] - OSError: [Errno 19] No such device
> ERROR tests/standard/fido2/pin/test_pin.py::TestPinAttempts::test_pin_attempts[123456789A] - OSError: [Errno 19] No such device
> ERROR tests/standard/fido2/pin/test_set_pin.py::TestSetPin::test_send_zero_length_pin_auth - OSError: [Errno 19] No such device
> ERROR tests/standard/fido2/user_presence/test_user_presence.py::TestUserPresence::test_user_presence_instructions - OSError: [Errno 19] No such device
> ERROR tests/standard/fido2/user_presence/test_user_presence.py::TestUserPresence::test_user_presence - OSError: [Errno 19] No such device
> ERROR tests/standard/fido2/user_presence/test_user_presence.py::TestUserPresence::test_no_user_presence - OSError: [Errno 19] No such device
> ERROR tests/standard/fido2/user_presence/test_user_presence.py::TestUserPresence::test_user_presence_option_false_on_get_assertion - OSError: [Errno 19] N...
> ERROR tests/standard/fido2/user_presence/test_user_presence.py::TestUserPresence::test_user_presence_option_false_on_make_credential - OSError: [Errno 19]...
> ERROR tests/standard/fido2/user_presence/test_user_presence.py::TestUserPresence::test_user_presence_permits_only_one_request - OSError: [Errno 19] No suc...
> ======================================= 23 failed, 92 passed, 8 skipped, 7 deselected, 17 errors in 106.16s (0:01:46) =======================================
> make[1]: *** [Makefile:19: standard-tests] Fehler 1
> make: *** [Makefile:20: fido2-test] Fehler 2
> kietzmann@kietzmann:/media/DATA/RIOT/tests/sys_fido2_ctap$ make fido2-test 
> make -C /media/DATA/RIOT/build/pkg/fido2_tests
> venv/bin/pytest --ignore=tests/standard/fido2v1 --ignore=tests/standard/fido2/extensions --ignore=tests/standard/u2f -k "not test_ctap1_interop and not test_rk_maximum_list_capacity_per_rp_nodisplay and not test_keep_alive" tests/standard/ -s
> ==================================================================== test session starts ====================================================================
> platform linux -- Python 3.8.5, pytest-6.2.4, py-1.10.0, pluggy-0.13.1
> rootdir: /media/DATA/RIOT/build/pkg/fido2_tests
> plugins: rerunfailures-10.0, ordering-0.6
> collected 147 items / 7 deselected / 140 selected                                                                                                           
> 
> tests/standard/connect_test.py --- HID ---
> [CtapHidDevice(/dev/hidraw0)]
> .
> tests/standard/fido2/test_get_assertion.py Resetting Authenticator...
> .......................Please reboot authenticator and hit enter
> 
> --- HID ---
> [CtapHidDevice(/dev/hidraw0)]
> .
> tests/standard/fido2/test_getinfo.py ...........
> tests/standard/fido2/test_make_credential.py Resetting Authenticator...
> .................................MC [b'123456789abcdef0123456789abcdef0', {'id': 'example.org', 'name': 'ExampleRP'}, {'id': b'N\x0f\x84\xc2\xff\xae\x11\x00\xb0d\x07^\xd1\xe0EWe\xcb\xd9\xdc\xf0[\x99\xd4<43\x18\xf7\x87\x94V\xd3\xce+\x13&\x14Y\xa4\xd1Ss\xc5\x8b', 'name': 'Jilleen Agnola Evey', 'icon': 'https://www.w3.org/TR/webauthn/', 'displayName': 'Displayed Jilleen Agnola Evey'}, [{'type': 'public-key', 'alg': -7}], None, None, {'unknown': False}, None, None, None, <tests.vendor.solo.utils.DeviceSelectCredential object at 0x7f94fa60c160>]
> .ed25519 is not supported.  Skip this test.
> .
> tests/standard/fido2/test_reset.py Resetting Authenticator...
> .
> tests/standard/fido2/test_reset_credential.py Resetting Authenticator...
> Resetting Authenticator...
> .
> tests/standard/fido2/test_resident_key.py Resetting Authenticator...
> .Please reboot authenticator and hit enter
> 
> --- HID ---
> [CtapHidDevice(/dev/hidraw0)]
> .Resetting Authenticator...
> .Please reboot authenticator and hit enter
> 
> --- HID ---
> [CtapHidDevice(/dev/hidraw0)]
> .Resetting Authenticator...
> Resetting Authenticator...
> .Resetting Authenticator...
> Resetting Authenticator...
> .Resetting Authenticator...
> ...
>             {'id': b'7\xc8\x05\xb9n\xc7\x7f\xee\xb7Zb\xf9\x89l"\xeb8QD\xbfG\xe4\xa3\x8a\x9e\xbd}\x98JP+2L\xc0\xf3\x81*Z\x0c\x90\xba\xa5\xa4h>}b\x08O\xce|', 'name': 'Morgana Athene Joscelin', 'icon': 'https://www.w3.org/TR/webauthn/', 'displayName': 'Displayed Morgana Athene Joscelin'}
>             b'123456789abcdef0123456789abcdef0'
>             {'id': 'unique-0.8474736579670072.com', 'name': 'Example'}
>             
> 
>             {'id': b'\xf7\x00\x97\xb74\x80\xa76ME\xaf\xebE\x81J\xda\xc2\x1a\xb9\x07\xa1A7@\xdb\xdcN\xad\n?\xfa\x12\xc9\xf4\xb2b', 'name': 'Marchelle Katie Willamina', 'icon': 'https://www.w3.org/TR/webauthn/', 'displayName': 'Displayed Marchelle Katie Willamina'}
>             b'123456789abcdef0123456789abcdef0'
>             {'id': 'unique-0.8474736579670072.com', 'name': 'Example'}
>             
> 
>             {'id': b"h\xe1\x9f'\x0f\xfc", 'name': 'Loise Christie Daniela', 'icon': 'https://www.w3.org/TR/webauthn/', 'displayName': 'Displayed Loise Christie Daniela'}
>             b'123456789abcdef0123456789abcdef0'
>             {'id': 'unique-0.8474736579670072.com', 'name': 'Example'}
>             
> .s.sResetting Authenticator...
> ...
>             {'id': b'_\x0e~\xaf\xf3\xd1\xcf}"\x13\x00\xa0\xce\x83', 'name': 'Jenifer Allie Danell', 'icon': 'https://www.w3.org/TR/webauthn/', 'displayName': 'Displayed Jenifer Allie Danell'}
>             b'123456789abcdef0123456789abcdef0'
>             {'id': 'unique-0.28421357863210517.com', 'name': 'Example'}
>             
> 
>             {'id': b'I\xd5\xd3\x81vg\xe0\x0c1\xf1\xb3G\xf0\xdc\x87H\x1cM\xdd\x1e\x9av\xf2Kd\xfe\x05', 'name': 'Fifine Margot Sandye', 'icon': 'https://www.w3.org/TR/webauthn/', 'displayName': 'Displayed Fifine Margot Sandye'}
>             b'123456789abcdef0123456789abcdef0'
>             {'id': 'unique-0.28421357863210517.com', 'name': 'Example'}
>             
> 
>             {'id': b'\xa4\xa9\xf9,\x10a)\x08\xf1\x04\xa4\xd3\xd9\xdf{\x11PT\xce', 'name': 'Bonnee Ellyn Antonia', 'icon': 'https://www.w3.org/TR/webauthn/', 'displayName': 'Displayed Bonnee Ellyn Antonia'}
>             b'123456789abcdef0123456789abcdef0'
>             {'id': 'unique-0.28421357863210517.com', 'name': 'Example'}
>             
> .s.sssResetting Authenticator...
> ...
> tests/standard/fido2/pin/test_lockout.py Resetting Authenticator...
> Please reboot authenticator and hit enter
> 
> --- HID ---
> [CtapHidDevice(/dev/hidraw0)]
> F
> tests/standard/fido2/pin/test_pin.py Resetting Authenticator...
> ........Resetting Authenticator...
> .Resetting Authenticator...
> Check there is 7 pin attempts left
> Check there is 6 pin attempts left
> Please reboot authenticator and hit enter
> 
> --- HID ---
> [CtapHidDevice(/dev/hidraw0)]
> F
> tests/standard/fido2/pin/test_set_pin.py Resetting Authenticator...
> .Resetting Authenticator...
> ....Resetting Authenticator...
> .
> tests/standard/fido2/user_presence/test_user_presence.py Resetting Authenticator...
> 
> 
> Starting User Presence (UP) tests.
> 
> Follow instructions.  You will have to give UP or not give UP to pass the tests.
> .ACTIVATE UP ONCE
> .DO NOT ACTIVATE UP
> FsDO NOT ACTIVATE UP
> .DO NOT ACTIVATE UP
> .ACTIVATE UP ONCE
> F
> tests/standard/transport/test_hid.py .b'\x11\x11\x11\x11\x11\x11\x11\x11\x0c\x00\x00\x00\x02\x00\x00\x00\r'
> ...Implements CBOR.
> .U2F not implemented.
> ............
> tests/standard/transport/test_nfc.py s
> 
> ========================================================================= FAILURES ==========================================================================
> _______________________________________________________________________ test_lockout ________________________________________________________________________
> 
> device = <tests.conftest.TestDevice object at 0x7f94f989a610>, resetDevice = <tests.conftest.TestDevice object at 0x7f94f989a610>
> 
>     @pytest.mark.skipif(
>         "trezor" in sys.argv, reason="ClientPin is not supported on Trezor."
>     )
>     def test_lockout(device, resetDevice):
>         pin = "TestPin"
>         device.client.pin_protocol.set_pin(pin)
>     
>         pin_token = device.client.pin_protocol.get_pin_token(pin)
>         req = FidoRequest(pin_token=pin_token)
>     
>         req.pin_auth = hmac_sha256(pin_token, req.cdh)[:16]
>     
>         for i in range(1, 10):
>             err = CtapError.ERR.PIN_INVALID
>             if i in (3, 6):
>                 err = CtapError.ERR.PIN_AUTH_BLOCKED
>             elif i >= 8:
>                 err = [CtapError.ERR.PIN_BLOCKED, CtapError.ERR.PIN_INVALID]
>     
>             with pytest.raises(CtapError) as e:
>                 device.sendPP("WrongPin")
> >           assert e.value.code == err or e.value.code in err
> E           TypeError: argument of type 'ERR' is not iterable
> 
> tests/standard/fido2/pin/test_lockout.py:30: TypeError
> _______________________________________________________ TestPinAttempts.test_pin_attempts[123456789A] _______________________________________________________
> 
> self = <tests.standard.fido2.pin.test_pin.TestPinAttempts object at 0x7f94f8d6be20>, device = <tests.conftest.TestDevice object at 0x7f94f989a610>
> SetPinRes = AttestationObject(fmt: 'packed', auth_data: AuthenticatorData(rp_id_hash: h'bfabc37432958b063360d3ad6461c9c4735ae7f8ed...xefj\xbe\x9eF\x106\xe4\x02 <q\xd8\xb5M\x8a\x9c\x025\xc7\x8a\xe7 G,h'\x1d\x94\xd9*\x13L\x96\x9eM=\xf8\xfd\x97\xdf\xf1"})
> 
>     @pytest.mark.skipif(
>         "trezor" in sys.argv, reason="ClientPin is not supported on Trezor."
>     )
>     def test_pin_attempts(self, device, SetPinRes):
>         # Flip 1 bit
>         pin = SetPinRes.PIN
>         pin_wrong = list(pin)
>         c = pin[len(pin) // 2]
>     
>         pin_wrong[len(pin) // 2] = chr(ord(c) ^ 1)
>         pin_wrong = "".join(pin_wrong)
>     
>         for i in range(1, 3):
>             with pytest.raises(CtapError) as e:
>                 device.sendPP(pin_wrong)
>             assert e.value.code == CtapError.ERR.PIN_INVALID
>     
>             print("Check there is %d pin attempts left" % (8 - i))
>             res = device.ctap2.client_pin(1, PinProtocolV1.CMD.GET_RETRIES)
>             assert res[3] == (8 - i)
>     
>         for i in range(1, 3):
>             with pytest.raises(CtapError) as e:
>                 device.sendPP(pin_wrong)
>             assert e.value.code == CtapError.ERR.PIN_AUTH_BLOCKED
>     
>         device.reboot()
>     
> >       reg = device.sendMC(*FidoRequest(SetPinRes, pin = pin).toMC())
> 
> tests/standard/fido2/pin/test_pin.py:189: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> tests/conftest.py:444: in sendMC
>     pin_token = self.client.pin_protocol.get_pin_token(pin)
> venv/lib/python3.8/site-packages/fido2/ctap2.py:905: in get_pin_token
>     key_agreement, shared_secret = self.get_shared_secret()
> venv/lib/python3.8/site-packages/fido2/ctap2.py:885: in get_shared_secret
>     resp = self.ctap.client_pin(
> venv/lib/python3.8/site-packages/fido2/ctap2.py:776: in client_pin
>     return self.send_cbor(
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> 
> self = <fido2.ctap2.CTAP2 object at 0x7f94f8d6b6d0>, cmd = <CMD.CLIENT_PIN: 6>, data = {1: 1, 2: <CMD.GET_KEY_AGREEMENT: 2>}, event = None
> parse = <function decode at 0x7f94f9b8e9d0>, on_keepalive = None
> 
>     def send_cbor(
>         self, cmd, data=None, event=None, parse=cbor.decode, on_keepalive=None
>     ):
>         """Sends a CBOR message to the device, and waits for a response.
>     
>         :param cmd: The command byte of the request.
>         :param data: The payload to send (to be CBOR encoded).
>         :param event: Optional threading.Event used to cancel the request.
>         :param parse: Function used to parse the binary response data, defaults
>             to parsing the CBOR.
>         :param on_keepalive: Optional function called when keep-alive is sent by
>             the authenticator.
>         :return: The result of calling the parse function on the response data
>             (defaults to the CBOR decoded value).
>         """
>         request = struct.pack(">B", cmd)
>         if data is not None:
>             request += cbor.encode(data)
>         response = self.device.call(CTAPHID.CBOR, request, event, on_keepalive)
>         status = six.indexbytes(response, 0)
>         if status != 0x00:
> >           raise CtapError(status)
> E           fido2.ctap.CtapError: CTAP error: 0x34 - PIN_AUTH_BLOCKED
> 
> venv/lib/python3.8/site-packages/fido2/ctap2.py:645: CtapError
> __________________________________________________________ TestUserPresence.test_no_user_presence ___________________________________________________________
> 
> self = <test_user_presence.TestUserPresence object at 0x7f94f8dd03a0>, device = <tests.conftest.TestDevice object at 0x7f94f989a610>
> MCRes = AttestationObject(fmt: 'packed', auth_data: AuthenticatorData(rp_id_hash: h'bfabc37432958b063360d3ad6461c9c4735ae7f8ed...\xd5\x9e1\x04\xe9\x84\x89\x02 ^\xcc\x15\xb0u\xbe\xc9,z\xa0\xbe9\xc9\xc5\x84B\xb2Jq\x1c\xcbe\x150P\x12\x7fZ\x80\xf7vi'})
> GARes = AssertionResponse(credential: {'id': b'\xcb\x07`\x1d*\x14\xaa\xe1k\xeb7DT\x90\xf4A+\r\xc3e\xecs\x16O\\\x87\x98\x01"\xf...b9f3c1f804d6d133f27d401713d05e16b4816cbc55f9814022100d814509fff921e79c2e7ed72a294cc3dcfade685e7114b1fa5807e6708d29be1')
> 
>     def test_no_user_presence(self, device, MCRes, GARes):
>         print("DO NOT ACTIVATE UP")
>         with pytest.raises(CtapError) as e:
>             with Timeout(2.0) as event:
> >               device.sendGA(
>                     *FidoRequest(GARes, timeout=event, on_keepalive=None).toGA()
>                 )
> E               Failed: DID NOT RAISE <class 'fido2.ctap.CtapError'>
> 
> tests/standard/fido2/user_presence/test_user_presence.py:34: Failed
> _______________________________________________ TestUserPresence.test_user_presence_permits_only_one_request ________________________________________________
> 
> self = <test_user_presence.TestUserPresence object at 0x7f94f8d45d90>, device = <tests.conftest.TestDevice object at 0x7f94f989a610>
> MCRes = AttestationObject(fmt: 'packed', auth_data: AuthenticatorData(rp_id_hash: h'bfabc37432958b063360d3ad6461c9c4735ae7f8ed...\xd5\x9e1\x04\xe9\x84\x89\x02 ^\xcc\x15\xb0u\xbe\xc9,z\xa0\xbe9\xc9\xc5\x84B\xb2Jq\x1c\xcbe\x150P\x12\x7fZ\x80\xf7vi'})
> GARes = AssertionResponse(credential: {'id': b'\xcb\x07`\x1d*\x14\xaa\xe1k\xeb7DT\x90\xf4A+\r\xc3e\xecs\x16O\\\x87\x98\x01"\xf...b9f3c1f804d6d133f27d401713d05e16b4816cbc55f9814022100d814509fff921e79c2e7ed72a294cc3dcfade685e7114b1fa5807e6708d29be1')
> 
>     def test_user_presence_permits_only_one_request(self, device, MCRes, GARes):
>         print("ACTIVATE UP ONCE")
>         device.sendGA(*FidoRequest(GARes).toGA())
>     
>         with pytest.raises(CtapError) as e:
>             with Timeout(1.0) as event:
> >               device.sendGA(
>                     *FidoRequest(GARes, timeout=event, on_keepalive=None).toGA()
>                 )
> E               Failed: DID NOT RAISE <class 'fido2.ctap.CtapError'>
> 
> tests/standard/fido2/user_presence/test_user_presence.py:80: Failed
> ================================================================== short test summary info ==================================================================
> FAILED tests/standard/fido2/pin/test_lockout.py::test_lockout - TypeError: argument of type 'ERR' is not iterable
> FAILED tests/standard/fido2/pin/test_pin.py::TestPinAttempts::test_pin_attempts[123456789A] - fido2.ctap.CtapError: CTAP error: 0x34 - PIN_AUTH_BLOCKED
> FAILED tests/standard/fido2/user_presence/test_user_presence.py::TestUserPresence::test_no_user_presence - Failed: DID NOT RAISE <class 'fido2.ctap.CtapEr...
> FAILED tests/standard/fido2/user_presence/test_user_presence.py::TestUserPresence::test_user_presence_permits_only_one_request - Failed: DID NOT RAISE <cl...
> ============================================ 4 failed, 128 passed, 8 skipped, 7 deselected in 114.88s (0:01:54) =============================================
> ```

This again looks like you didn't wait long enough for the connection between authenticator and host computer to be reestablished. When I run the tests I get following failures still:

```
FAILED tests/standard/fido2/pin/test_lockout.py::test_lockout - ValueError: Extraneous data
FAILED tests/standard/fido2/user_presence/test_user_presence.py::TestUserPresence::test_no_user_presence - Failed: DID NOT RAISE <class 'fido2.ctap.CtapError'>
FAILED tests/standard/fido2/user_presence/test_user_presence.py::TestUserPresence::test_user_presence_permits_only_one_request - Failed: DID NOT RAISE <class 'fido2.ctap.CtapError'>
```
The user presence errors occur since I have disabled UP for testing. I am still thinking about what to do about these tests. Probably need to test with UP turned on. This would be a real pain though. 
The extraneous data error is the one I still can not explain and which I think is originating in the USB layer.

Reply:
There is no encryption happening here. It is just the key that is being created.

Reply:
No it is not expected. The CTAP spec just mentions the possibility of storing a credential on a server in encrypted form. It doesn't define how. 

But to me it makes sense to initialize a new key in the reset method because this invalidates all credentials that are stored remotely since the Authenticator won't be able to decrypt them anymore with the new key. This to me is also part of the "factory default state".

I could also simply invalidate the old key (e.g. all 0) and initialize a new key the next time a credential needs to be stored remotely on a server.